### PR TITLE
Add support for SaxonHE for xsl processing

### PR DESCRIPTION
**Pull request by djarek**
Will be used in generation of Beast reference.

### CODE REVIEW

**Comment by swatanabe** [comment]
AMDG

- C:\Boost\bin seems like an odd place to search for java.
- Use common.get-invocation-command instead of rolling your own.
- You should probably search common installation locations
  for the .jar, e.g. /usr/share/java/saxon/Saxon-HE.jar.
- There's no need for extra indirection in the saxonhe rule:
  actions saxonhe
  {
    "$(.java_exe)" -jar $(.saxonhe_jar) -o:$(<) -s:$(>[1]) -xsl:$(>[2])
  }

In Christ,
Steven Watanabe

**Comment by djarek** [comment]
@swatanabe 
Most of this was copy-pasted from xsltproc.jam's implementation. Should I add a default path to search the jar for on Windows?

**Comment by swatanabe** [comment]
AMDG

On 12/19/2018 06:39 PM, Damian Jarek wrote:
> @swatanabe 
> Most of this was copy-pasted from xsltproc.jam's implementation.

The extra complexity in xsltproc is because it's doing more.

> Should I add a default path to search the jar for on Windows?
> 

Is there a standard install location on Windows?

In Christ,
Steven Watanabe

**Comment by djarek** [comment]
Hmm, couldn't find anything about a standard location for *.jar files on Windows, so I presume there is none.

**Comment by swatanabe** [comment]
AMDG

On 12/20/2018 02:07 PM, Damian Jarek wrote:
> Hmm, couldn't find anything about a standard location for *.jar files on Windows, so I presume there is none.
> 

That's what it looks like to me, too.  There's
no installer.  You just unzip the archive
somewhere.  You might make it so that it works
as long as the .jar is in CLASSPATH, though.

In Christ,
Steven Watanabe

**Comment by vinniefalco** [comment]
What else is needed from this pull request in order to merge?

**Comment by vinniefalco** [comment]
So...what's the hold up here?

**Comment by swatanabe** [comment]
AMDG

On 1/19/19 11:04 AM, Vinnie Falco wrote:
> So...what's the hold up here?
> 

Christmas got in the way, and then I forgot about it.

Anyway,
- get-invocation-command should only be used for
  executables, not for .jars, since it searches PATH.
- The default install location on Fedora is /usr/share/java/saxon/,
  so it should probably be listed in addition to /usr/share/java.
- The action needs to make sure that all file paths are quoted.

In Christ,
Steven Watanabe

**Comment by djarek** [comment]
What should I use instead of `get-invocation-command`?

**Comment by swatanabe** [comment]
AMDG

On 2/1/19 7:11 PM, Damian Jarek wrote:
> What should I use instead of `get-invocation-command`?
> 

Just use path.glob like I think you used
in the original PR.  Executables need special,
platform specific, handling, which is why there's
a special rule for them.  .jars are simpler,
so a basic glob is good enough.

In Christ,
Steven Watanabe

**Comment by vinniefalco** [comment]
Just out of curiosity, what would that look like exactly if it was formed into a complete, working pull request?

**Comment by djarek** [comment]
Changed the init rule so now it uses a regular glob to find the saxon jars.

File: src/tools/saxonhe.jam [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+rule init ( saxonhe_jar ? : java_exe ? )
+{
+    .java_exe = [ common.get-invocation-command saxonhe : java : $(java_exe) :  ] ;
+    if $(saxonhe_jar)
+    {
+        .saxonhe_jar = $(saxonhe_jar) ;
+    }
+    else
+    {
+        local jar = [ GLOB $(path) "/usr/share/java/saxon/" "/usr/share/java/" : Saxon-HE.jar ] ;
```
[the remark this answers is not in the document]
Comment (djarek) [reply src/tools/saxonhe.jam]: Hmm, no idea where I got that from, anyway, removed the use of `$path`.

**Comment by swatanabe** [comment]
AMDG

The init rule looks good to me now.  You just
need to quote all paths in the action.

In Christ,
Steven Watanabe

File: src/tools/saxonhe.jam [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright (c) 2018 Damian Jarek (damian dot jarek93 at gmail dot com)
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
```
Comment (vinniefalco) [inline src/tools/saxonhe.jam]: I would throw another pound sign on its own line here but that's just me

**Comment by djarek** [comment]
All paths in the action are now quoted.

**Comment by swatanabe** [comment]
AMDG

On 2/13/19 3:01 PM, Damian Jarek wrote:
> All paths in the action are now quoted.
> 

Looks good to me.

In Christ,
Steven Watanabe

**Comment by vinniefalco** [comment]
**HALLELUJAH** !!!!

**Pull request merged**